sys/ztimer/periodic.c

```diff
@@ -54,32 +54,28 @@ static void _ztimer_periodic_callback(void *arg)
 }
 
 void ztimer_periodic_init(ztimer_clock_t *clock, ztimer_periodic_t *timer,
-                          bool (*callback)(
-                              void *), void *arg, uint32_t interval)
+                          bool (*callback)(void *), void *arg, uint32_t interval)
 {
-    ztimer_remove(clock, &timer->timer);
-    *timer =
-        (ztimer_periodic_t){ .clock = clock, .interval = interval,
-                             .callback = callback, .arg = arg,
-                             .timer = {
-                                 .callback = _ztimer_periodic_callback,
-                                 .arg = timer
-                             } };
+    /* check if this is a reinit, ensure timer is stopped in case */
+    if (timer->timer.callback == _ztimer_periodic_callback) {
+        ztimer_periodic_stop(timer);
+    }
+    *timer = (ztimer_periodic_t){
+        .clock = clock, .interval = interval,
+        .callback = callback, .arg = arg,
+        .timer = {
+            .callback = _ztimer_periodic_callback,
+            .arg = timer
+        }
+    };
 }
 
 void ztimer_periodic_start(ztimer_periodic_t *timer)
 {
-    ztimer_acquire(timer->clock);
-
-    uint32_t now = ztimer_now(timer->clock);
-
-    timer->last = now;
-    _ztimer_periodic_reset(timer, now);
+    timer->last = ztimer_set(timer->clock, &timer->timer, timer->interval) + timer->interval;
 }
 
 void ztimer_periodic_stop(ztimer_periodic_t *timer)
 {
     ztimer_remove(timer->clock, &timer->timer);
```

Comment:
Maybe set `.timer.callback = NULL` here?

Reply:
We can, but there is no reason to, since we do not need to, since we do not need to track if the timer is running when we dont need to aquire since we are not using ztimer_now in start any more by setting the ztimer directly and using its return value

Reply:
But, `ztimer_periodic_init()` calls `ztimer_periodic_stop()` if `.timer.callback == _ztimer_periodic_callback`. So if the user does `stop()`, `init()`, then `stop()` gets called by `init()` again uselessly on a stopped timer, causing a full timer list iteration. Setting the callback to sth else than `_ztimer_periodic_callback` here avoids that.

Reply:
I wouldn't worry about `ztimer_periodic_init()` being on any kind of fast path, that shouldn't be something that gets called often.

Reply:
Well, preventing interrupts from getting disabled might help hot paths elsewhere.
But I see that's a potentially non-trivial optimization, fine with me to leave out.